### ISSUE TITLE
Intermittent test failures

### ISSUE THREAD

**Issue by matthew-white** [getodk/central-backend#588]
Tests often pass locally for me, but there are sometimes one or more failures. The failures aren't consistent: I've seen a few different errors, and even when it's the same error, it's often a different test that fails. It sounds like @ktuite is seeing some of these same errors, so I thought I'd create an issue to track them. Rather than creating an issue for each error, I thought I'd create a single issue to track the different errors we've seen. (We can always spin one off into its own issue if it starts a discussion, but starting things here seems like it might work well.) Feel free to comment if you encounter an error not listed here, to :+1: an error that you've seen, or to add more information to any of the comments below.

**Comment by matthew-white** [comment]
@ktuite [mentioned](https://github.com/getodk/central-backend/pull/583#issuecomment-1230701615):

> I occasionally see this ` Error: end of central directory record signature not found` error on submission zip tests. Not all the time, not always the same test. But the tests all involve `yauzl`.

```
  1) api: /forms/:id/submissions
       [draft] .csv.zip
         should return draft submissions:
     Error: end of central directory record signature not found
      at /Users/ktuite/Desktop/code/odk/central-backend/node_modules/yauzl/index.js:183:14
      at Immediate.<anonymous> (node_modules/yauzl/index.js:615:38)
      at processImmediate (internal/timers.js:464:21)
```

I've seen a similar error:

```
  2) managed encryption
       end-to-end
         should decrypt over cookie auth with passphrases provided via url-encoded POST body:
     Error: end of central directory record signature not found
      at node_modules/yauzl/index.js:183:14
      at node_modules/yauzl/index.js:622:5
      at node_modules/fd-slicer/index.js:32:7
      at FSReqCallback.wrapper [as oncomplete] (fs.js:561:5)
```

Also in these tests:

```
  1) zipPart streamer
       should close the archive only after parts are finalized:

  2) .csv.zip briefcase output @slow
       should output all rows:

  1) api: /forms/:id/submissions
       .csv.zip GET
         versioning
           should return original instanceId and latest attached audit log when instanceId deprecated with new audit log:
```

I see this error maybe 25% of the time that I run tests.

I've seen this error mentioned on the forum: https://forum.getodk.org/t/error-end-of-central-directory-record-signature-not-found/36182

**Comment by matthew-white** [comment]
I see this error maybe 10% of the time that I run tests:

```
     Error: socket hang up
      at connResetException (internal/errors.js:628:14)
      at Socket.socketOnEnd (_http_client.js:499:23)
      at Socket.emit (domain.js:470:12)
      at endReadableNT (internal/streams/readable.js:1317:12)
      at processTicksAndRejections (internal/process/task_queues.js:82:21)
```

I've seen the error in these tests:

```
  1) managed encryption
       end-to-end
         should handle mixed [plaintext/encrypted] attachments (not decrypting):

  1) analytics task queries
       submission metrics
         should calculate submissions that have comments:
```

**Comment by matthew-white** [comment]
I see this error maybe 10% of the time that I run tests:

```
     Error: Parse Error: Expected HTTP/
      at Socket.socketOnData (_http_client.js:515:22)
      at Socket.emit (domain.js:470:12)
      at addChunk (internal/streams/readable.js:290:12)
      at readableAddChunk (internal/streams/readable.js:265:9)
      at Socket.Readable.push (internal/streams/readable.js:204:10)
      at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
```

I've seen the error in these tests:

```
  1) api: /users
       /reset/initiate POST
         should send a specific email if an account existed but was deleted:

  1) api: /submission
       POST
         should store the correct formdef and actor ids:

  1) api: /forms/:id.svc
       /draft.svc
         /Submissions.xyz
           should not return results from the published form:
```

**Comment by matthew-white** [comment]
I see this error maybe 40% of the time that I run tests:

```
     TypeError: Cannot read property 'status' of undefined
      at Test._assertStatus (node_modules/supertest/lib/test.js:263:11)
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)
      at Test.assert (node_modules/supertest/lib/test.js:171:18)
      at Server.assert (node_modules/supertest/lib/test.js:131:12)
      at Server.emit (domain.js:470:12)
      at emitCloseNT (net.js:1661:8)
      at processTicksAndRejections (internal/process/task_queues.js:81:21)
```

I've seen that error in these tests:

```
  1) api: /forms/:id/submissions
       /:instanceId PATCH
         should reject if the user cannot update:

  1) api: /forms/:id/submissions
       /keys GET
         should return a self-managed key if it is used:

  1) /audits
       GET
         should filter out submission and backup events given action=nonverbose:

  2) project viewer role
       should be able to list form submissions:

  1) project viewer role
       should be able to fetch a submission attachment:

  1) api: /users
       /users/:id PATCH
         should update only the allowed fields:

  1) api: /projects
       GET
         should return all projects for an administrator:

  1) api: /projects/:id/forms (drafts)
       /:id/draft
         POST
           should deal with xlsx itemsets:

  1) api: /projects/:id/forms (versions)
       /:id/versions
         /:version
           /manifest GET
             should return a manifest:

  1) api: /assignments
       /:roleId/:actorId
         POST
           should create an assignment by role system id:

  1) api: /submission
       POST
         versioning
           should set the submission review state to edited:
```

**Comment by alxndrsn** [comment]
@matthew-white does https://github.com/getodk/central-backend/issues/499 issue look similar to your `_assertStatus` failures?  I think the logs there give a bit more clue as to what could be failing.

> I see this error maybe 40% of the time that I run tests:

Does the failure rate decrease if you increase the test mocha default test timeout?

**Comment by matthew-white** [comment]
I've seen this test failure multiple times, but it's rare:

```
  1) api: /forms/:id/submissions
       .csv.zip GET
         should return worker-processed consolidated client audit log attachments:

      AssertionError: expected 'instance ID,event,node,start,end,latitude,longitude,accuracy,old-value,new-value\none,e,/data/e,2000-01-01T00:11,,,,,hh,ii\none,d,/data/d,2000-01-01T00:10,,10,11,12,gg,\none,c,/data/c,2000-01-01T00:03,2000-01-01T00:04,7,8,9,ee,ff\none,b,/data/b,2000-01-01T00:02,2000-01-01T00:03,4,5,6,cc,dd\none,a,/data/a,2000-01-01T00:01,2000-01-01T00:02,1,2,3,aa,bb\ntwo,h,/data/h,2000-01-01T00:06,2000-01-01T00:07,-5,-6,,ee,ff\ntwo,g,/data/g,2000-01-01T00:05,2000-01-01T00:06,-3,-4,,cc,dd\ntwo,f,/data/f,2000-01-01T00:04,2000-01-01T00:05,-1,-2,,aa,bb\n' to be 'instance ID,event,node,start,end,latitude,longitude,accuracy,old-value,new-value\none,a,/data/a,2000-01-01T00:01,2000-01-01T00:02,1,2,3,aa,bb\none,b,/data/b,2000-01-01T00:02,2000-01-01T00:03,4,5,6,cc,dd\none,c,/data/c,2000-01-01T00:03,2000-01-01T00:04,7,8,9,ee,ff\none,d,/data/d,2000-01-01T00:10,,10,11,12,gg,\none,e,/data/e,2000-01-01T00:11,,,,,hh,ii\ntwo,f,/data/f,2000-01-01T00:04,2000-01-01T00:05,-1,-2,,aa,bb\ntwo,g,/data/g,2000-01-01T00:05,2000-01-01T00:06,-3,-4,,cc,dd\ntwo,h,/data/h,2000-01-01T00:06,2000-01-01T00:07,-5,-6,,ee,ff\n'
      + expected - actual

       instance ID,event,node,start,end,latitude,longitude,accuracy,old-value,new-value
      +one,a,/data/a,2000-01-01T00:01,2000-01-01T00:02,1,2,3,aa,bb
      +one,b,/data/b,2000-01-01T00:02,2000-01-01T00:03,4,5,6,cc,dd
      +one,c,/data/c,2000-01-01T00:03,2000-01-01T00:04,7,8,9,ee,ff
      +one,d,/data/d,2000-01-01T00:10,,10,11,12,gg,
       one,e,/data/e,2000-01-01T00:11,,,,,hh,ii
      -one,d,/data/d,2000-01-01T00:10,,10,11,12,gg,
      -one,c,/data/c,2000-01-01T00:03,2000-01-01T00:04,7,8,9,ee,ff
      -one,b,/data/b,2000-01-01T00:02,2000-01-01T00:03,4,5,6,cc,dd
      -one,a,/data/a,2000-01-01T00:01,2000-01-01T00:02,1,2,3,aa,bb
      +two,f,/data/f,2000-01-01T00:04,2000-01-01T00:05,-1,-2,,aa,bb
      +two,g,/data/g,2000-01-01T00:05,2000-01-01T00:06,-3,-4,,cc,dd
       two,h,/data/h,2000-01-01T00:06,2000-01-01T00:07,-5,-6,,ee,ff
      -two,g,/data/g,2000-01-01T00:05,2000-01-01T00:06,-3,-4,,cc,dd
      -two,f,/data/f,2000-01-01T00:04,2000-01-01T00:05,-1,-2,,aa,bb

      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at test/integration/api/submissions.js:1799:51
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at async Object.transaction (node_modules/slonik/dist/src/connectionMethods/transaction.js:22:24)
      at async Object.createConnection (node_modules/slonik/dist/src/factories/createConnection.js:96:18)
```

**Comment by matthew-white** [comment]
> does #499 issue look similar to your `_assertStatus` failures? I think the logs there give a bit more clue as to what could be failing.

I think they're similar insomuch as they're both failing at an `.expect()` call that's making an assertion about the status code. But in #499, it's the wrong status code, whereas above, it's more like there isn't even a response to check the status. Also, I didn't see additional logging like #499.

visionmedia/supertest#352 has some related discussion. Maybe we could try updating SuperTest? Though it doesn't seem like the version we're on should be having this issue.

> Does the failure rate decrease if you increase the test mocha default test timeout?

For some reason, I'm now seeing test failures at a much lower rate. I'm not seeing some of the errors that I did yesterday. I'm still seeing `Error: end of central directory record signature not found`, and I'm also seeing a [new error](https://github.com/getodk/central-backend/issues/588#issuecomment-1231882947). I'm going to keep trying over the next few days to see if there are changes.

**Comment by matthew-white** [comment]
I've seen this test failure just once so far:

```
  1) api: /forms/:id/submissions
       .csv.zip GET
         should split select multiples and filter given both options:
     Uncaught TypeError: Cannot read property 'call' of undefined
      at Test.assert (node_modules/supertest/lib/test.js:179:6)
      at Server.assert (node_modules/supertest/lib/test.js:131:12)
      at Server.emit (domain.js:470:12)
      at emitCloseNT (net.js:1661:8)
      at processTicksAndRejections (internal/process/task_queues.js:81:21)
```

It's another SuperTest error, so may be related to `TypeError: Cannot read property 'status' of undefined` [above](https://github.com/getodk/central-backend/issues/588#issuecomment-1230768278)?

After this test failure, 190 other tests timed out.

**Comment by matthew-white** [comment]
I've seen this test failure just once so far:

```
  1) api: /projects/:id/assignments/forms
       GET
         should return all assignments on forms:
     Error: ECONNREFUSED: Connection refused
      at Test.assert (node_modules/supertest/lib/test.js:164:13)
      at assert (node_modules/supertest/lib/test.js:131:12)
      at node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at ClientRequest.<anonymous> (node_modules/superagent/lib/node/index.js:647:10)
      at ClientRequest.emit (domain.js:470:12)
      at Socket.socketErrorListener (_http_client.js:475:9)
      at Socket.emit (domain.js:470:12)
      at emitErrorNT (internal/streams/destroy.js:106:8)
      at emitErrorCloseNT (internal/streams/destroy.js:74:3)
      at processTicksAndRejections (internal/process/task_queues.js:82:21)
```

**Comment by matthew-white** [comment]
I wrote a little Bash script to run `make test` multiple times, displaying just the summary from each run:

```bash
testloop() {
	local count=${1:-10}
	touch testloop.txt
	local temp=$(mktemp)
	local i line
	for i in $(seq $count); do
		make test | tee $temp
		line=$(grep -n -m 1 passing $temp | cut -d ':' -f 1)
		tail -n +$line $temp >> testloop.txt
	done
}
testloop "$@"
```

@alxndrsn, do you see any of these errors if you run tests say 25 times?

My current theory is that `Error: end of central directory record signature not found` and the `AssertionError` for `should return worker-processed consolidated client audit log attachments` will show up on a variety of machines, whereas for many of the others, there has to be some condition present on the machine related to connections.

**Comment by alxndrsn** [comment]
Seen:

# 1.

I think this is fine:
```js
  1) "before all" hook: initialize in "{root}":
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)
```

# 2.

```js
  1) "before all" hook: initialize in "{root}":
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)
```

# 3. 
```js
  1) managed encryption
       end-to-end
         should reject with a reasonable message given incorrect passphrase:
     Error: expected 400 "Bad Request", got 500 "Internal Server Error"
      at Test._assertStatus (node_modules/supertest/lib/test.js:266:12)
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)
      at Test.assert (node_modules/supertest/lib/test.js:171:18)
      at Server.assert (node_modules/supertest/lib/test.js:131:12)
      at Server.emit (domain.js:475:12)
      at emitCloseNT (net.js:1677:8)
      at processTicksAndRejections (internal/process/task_queues.js:81:21)
```

**Comment by matthew-white** [comment]
I've also seen that `"before all" hook: initialize in "{root}"` timeout (just once over the past few days). I agree that it seems fine.

I ran my Bash script on CircleCI, running `make test` 100 times (over [two](https://app.circleci.com/pipelines/github/getodk/central-backend/943/workflows/ae2d8b35-5f61-4ca1-afcf-8d6a2f99e40a/jobs/1695/artifacts) [builds](https://app.circleci.com/pipelines/github/getodk/central-backend/943/workflows/223477a5-ee43-4c3a-8964-1e37893a206a/jobs/1696/artifacts)). Mostly the only issue was timeouts, but `Error: end of central directory record signature not found` and the `AssertionError` for `should return worker-processed consolidated client audit log attachments` each came up once. Since we're seeing those on a variety of machines, and they don't seem connected to the other errors I've seen, I think I'll go ahead and file individual issues for those. I'll probably stop looking into those errors for now, but anyone is welcome to keep digging.

I'm continuing to see other errors, but only sometimes. One thing I tried doing is to update SuperTest to the latest version. `TypeError: Cannot read property 'status' of undefined` went away, so I think updating had an effect. After updating, I think I'm seeing a greater number of `Error: socket hang up` and `Error: Parse Error: Expected HTTP/`. My thinking is that those are the actual errors and that the latest version of SuperTest has better error handling that will surface the underlying error instead of failing during an assertion. At some point, I'll push a PR to upgrade SuperTest, since I think that's potentially helpful. I'll also try to look more into why I'm seeing so many socket hang-ups. (Let me know if anyone has ideas there!)

**Comment by alxndrsn** [comment]
Here's a new one:

```
1 failing

  1) task: fs
       encrypted archives
         should round-trip successfully:
     AssertionError: expected Array [ 'one' ] to contain Array [ 'one', 'two' ]
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value [as containDeep] (node_modules/should/cjs/should.js:356:19)
      at /home/circleci/repo/test/integration/task/fs.js:33:20
      at async Object.transaction (node_modules/slonik/dist/src/connectionMethods/transaction.js:22:24)
      at async Object.createConnection (node_modules/slonik/dist/src/factories/createConnection.js:97:18)
```

https://app.circleci.com/pipelines/github/getodk/central-backend/1156/workflows/5b3e5294-0fd2-4f77-a495-233716b88306/jobs/1940

**Comment by alxndrsn** [comment]
Seen with current `master` (d8181acab5dd000eac6d60d0df976f1e91701147):

```
  1097 passing (3m)
  14 pending
  3 failing

  1) datbase migrations: removing default project
       should put old forms into project:
     Error: Timeout of 4000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/user/workspaces/odk/odk-central-backend/test/integration/other/migrations.js)
      at listOnTimeout (node:internal/timers:559:17)
      at processTimers (node:internal/timers:502:7)

  2) datbase migrations: intermediate form schema
       should test migration:
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/user/workspaces/odk/odk-central-backend/test/integration/other/migrations.js)
      at listOnTimeout (node:internal/timers:559:17)
      at processTimers (node:internal/timers:502:7)

  3) "after each" hook for "should test migration":
     error: drop owned by current_user - deadlock detected
      at Parser.parseErrorMessage (node_modules/pg-protocol/dist/parser.js:287:98)
      at Parser.handlePacket (node_modules/pg-protocol/dist/parser.js:126:29)
      at Parser.parse (node_modules/pg-protocol/dist/parser.js:39:38)
      at Socket.<anonymous> (node_modules/pg-protocol/dist/index.js:11:42)
      at Socket.emit (node:events:513:28)
      at Socket.emit (node:domain:489:12)
      at addChunk (node:internal/streams/readable:315:12)
      at readableAddChunk (node:internal/streams/readable:289:9)
      at Socket.Readable.push (node:internal/streams/readable:228:10)
      at TCP.onStreamRead (node:internal/stream_base_commons:190:23)



make: *** [Makefile:32: test] Error 3
```

Adding `--timeout=0` to the `mocha` invocation in the `test` `make` target changes the error slightly:

```
  datbase migrations: intermediate form schema
    1) should test migration
    2) "after each" hook for "should test migration"


  1098 passing (2m)
  14 pending
  2 failing

  1) datbase migrations: intermediate form schema
       should test migration:
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/user/workspaces/odk/odk-central-backend/test/integration/other/migrations.js)
      at listOnTimeout (node:internal/timers:559:17)
      at processTimers (node:internal/timers:502:7)

  2) "after each" hook for "should test migration":
     error: drop owned by current_user - deadlock detected
      at Parser.parseErrorMessage (node_modules/pg-protocol/dist/parser.js:287:98)
      at Parser.handlePacket (node_modules/pg-protocol/dist/parser.js:126:29)
      at Parser.parse (node_modules/pg-protocol/dist/parser.js:39:38)
      at Socket.<anonymous> (node_modules/pg-protocol/dist/index.js:11:42)
      at Socket.emit (node:events:513:28)
      at Socket.emit (node:domain:489:12)
      at addChunk (node:internal/streams/readable:315:12)
      at readableAddChunk (node:internal/streams/readable:289:9)
      at Socket.Readable.push (node:internal/streams/readable:228:10)
      at TCP.onStreamRead (node:internal/stream_base_commons:190:23)



make: *** [Makefile:32: test] Error 2
```

Modifying all explicit [`this.timeout(...)`](https://mochajs.org/#timeouts) calls to `this.timeout(0)` allows the tests to pass, but is inconvenient as it requires constant change management when working with `git`.

This has been discussed previously at https://github.com/getodk/central-backend/issues/532.

Another option that works on my machine is doubling all the explicit timeouts.  I've opened a PR for this at https://github.com/getodk/central-backend/pull/847.

**Comment by alxndrsn** [comment]
Supertest updated in https://github.com/getodk/central-backend/pull/899

**Comment by alxndrsn** [comment]
Seeing intermittent faliures with auth(?) for OIDC tests, e.g.:

### [`oidc-integration-test`](https://github.com/getodk/central-backend/actions/runs/6348394324/job/17244924010#logs)

```
2023-09-29T07:28:39.4110341Z ::ffff:127.0.0.1 - - [29/Sep/2023:07:28:39 +0000] "GET /v1/oidc/login HTTP/1.1" 307 331
2023-09-29T07:28:39.4361117Z }attempted to log Sentry exception in development:
2023-09-29T07:28:39.4365708Z OPError: expected 200 OK, got: 408 Request Timeout
2023-09-29T07:28:39.4366590Z     at processResponse (/home/runner/work/central-backend/central-backend/node_modules/openid-client/lib/helpers/process_response.js:41:11)
2023-09-29T07:28:39.4367336Z     at Client.userinfo (/home/runner/work/central-backend/central-backend/node_modules/openid-client/lib/client.js:1237:18)
2023-09-29T07:28:39.4367857Z     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2023-09-29T07:28:39.4368425Z     at async /home/runner/work/central-backend/central-backend/lib/resources/oidc.js:99:24 {
2023-09-29T07:28:39.4368890Z   error: 'expected 200 OK, got: 408 Request Timeout'
2023-09-29T07:28:39.4369153Z }
2023-09-29T07:28:39.4383034Z ::ffff:127.0.0.1 - - [29/Sep/2023:07:28:39 +0000] "GET /v1/oidc/callback?code=ItcuKk91nQamjq81DjbXSly4kpezOlYJcjfovU1Wt35&iss=http%3A%2F%2Flocalhost%3A9898 HTTP/1.1" 303 87
2023-09-29T07:28:39.4401946Z         1) should update the label of an entity
```

**Comment by matthew-white** [comment]
> This looks like a new one:
> 
> ```
>   1) api: /projects/:id/forms (drafts)
>        /:id/draft
>          POST
>            should stop waiting for Enketo after 0.5 seconds @slow:
>      AssertionError: expected '::abcdefgh' to not exist
> ```

As of #1001, this test should no longer fail intermittently. I'll mark your comment as resolved, but let me know if you see the failure again.

**Comment by matthew-white** [comment]
@alxndrsn, I feel like we're doing better with intermittent test failures than when I first filed this issue. Glancing over this issue, it's unclear to me which failures are still an issue and which have been resolved (e.g., by the update to SuperTest). What do you think about us closing this issue, then filing an individual issue for each new test failure we see in the future?

**Comment by matthew-white** [comment]
If there are any test failures listed here that you know are still an issue or that you don't want to miss, I'd be happy to file issues for them.

**Comment by alxndrsn** [comment]
> If there are any test failures listed here that you know are still an issue or that you don't want to miss, I'd be happy to file issues for them.

I think it's inevitable that this issue is stale, so I think we can just open new tickets for any failures we observe in future.
